crates/bevy_ecs/src/schedule_v3/mod.rs

```diff
@@ -0,0 +1,236 @@
+        assert!(matches!(result, Err(BuildError::DependencyCycle)));
+    }
+
+    #[test]
+    #[should_panic]
+    fn hierarchy_loop() {
+        let mut schedule = Schedule::new();
+        schedule.configure_set(TestSet::X.in_set(TestSet::X));
+    }
+
+    #[test]
+    fn hierarchy_cycle() {
+        let mut world = World::new();
+        let mut schedule = Schedule::new();
+
+        schedule.configure_set(TestSet::A.in_set(TestSet::B));
+        schedule.configure_set(TestSet::B.in_set(TestSet::A));
+
+        let result = schedule.initialize(&mut world);
+        assert!(matches!(result, Err(BuildError::HierarchyCycle)));
+    }
+
+    #[test]
+    fn system_type_set_ambiguity() {
+        // Define some systems.
+        fn foo() {}
+        fn bar() {}
+
+        let mut world = World::new();
+        let mut schedule = Schedule::new();
+
+        // Schedule `bar` to run after `foo`.
+        schedule.add_system(foo);
+        schedule.add_system(bar.after(foo));
+
+        // There's only one `foo`, so it's fine.
+        let result = schedule.initialize(&mut world);
+        assert!(result.is_ok());
+
+        // Schedule another `foo`.
+        schedule.add_system(foo);
+
+        // When there are multiple instances of `foo`, dependencies on
+        // `foo` are no longer allowed. Too much ambiguity.
+        let result = schedule.initialize(&mut world);
+        assert!(matches!(result, Err(BuildError::SystemTypeSetAmbiguity(_))));
+
+        // same goes for `ambiguous_with`
+        let mut schedule = Schedule::new();
+        schedule.add_system(foo);
+        schedule.add_system(bar.ambiguous_with(foo));
+        let result = schedule.initialize(&mut world);
+        assert!(result.is_ok());
+        schedule.add_system(foo);
+        let result = schedule.initialize(&mut world);
+        assert!(matches!(result, Err(BuildError::SystemTypeSetAmbiguity(_))));
+    }
+
+    #[test]
+    #[should_panic]
+    fn in_system_type_set() {
+        fn foo() {}
+        fn bar() {}
+
+        let mut schedule = Schedule::new();
+        schedule.add_system(foo.in_set(bar.into_system_set()));
+    }
+
+    #[test]
+    #[should_panic]
+    fn configure_system_type_set() {
+        fn foo() {}
+        let mut schedule = Schedule::new();
+        schedule.configure_set(foo.into_system_set());
+    }
+
+    #[test]
+    fn hierarchy_conflict() {
+        let mut world = World::new();
+        let mut schedule = Schedule::new();
+
+        // Add `A`.
+        schedule.configure_set(TestSet::A);
+
+        // Add `B` as child of `A`.
+        schedule.configure_set(TestSet::B.in_set(TestSet::A));
+
+        // Add `X` as child of both `A` and `B`.
+        schedule.configure_set(TestSet::X.in_set(TestSet::A).in_set(TestSet::B));
+
+        // `X` cannot be the `A`'s child and grandchild at the same time.
+        let result = schedule.initialize(&mut world);
+        assert!(matches!(result, Err(BuildError::HierarchyConflict)));
```

Comment:
I'm still conflicted over the usefulness of this being an error. Won't block on it, but calling it out.

Reply:
This is definitely an error. We can "fix" it automatically by replacing the hierarchy with its transitive reduction, but if you see this, you're doing something weird and should probably fix it manually.

Reply:
Made this opt-in.